rfcs/0019-generalized-transactions.md

```diff
@@ -0,0 +1,190 @@
+# Generalized accounts and their specialization to normal accounts and tokens
+
+This is a proposal for generalized accounts which is conceptually simple and
+specializes to the case of normal Coda accounts as well as other token accounts.
+
+The idea is that each account will contain a state as well as a program which approves
+or rejects updates to that state, given also the state of other accounts involved in a transaction.
+I will use type-theory pseduocode to express the idea.
+
+The general framework is the following. We'll have a limited "programming language" 
+which I'll described as
+
+- a type family `predicate : (witness : type) (input : type) -> type`
+- a function `eval : forall (w a : type), (p : predicate w a) -> global_state -> w -> a -> bool`.
+
+where `global_state` is TBD, but should at least contain the current consensus state and a counter.
+
+Accounts will be described by a type family
+```ocaml
+let product : list type -> type = fold ( * ) unit
+
+type account_data (s : type) =
+  { state : s
+  ; balance : amount
```

Comment:
Emre mentioned that it would be beneficial to pay fees in any token, rather than just coda. Eg. If i only care about Echotokens, it would be nice to not have to know anything about Coda.

I also would like to understand the shape of accounts a bit more -- I can't fully decipher the types in ocaml yet, but as far as I can tell, the `account_data` would contain a `balance` in Coda, and then the nested `state` would have the token `balance`?

Reply:
If fees must be paid in Coda currency then Coda currency has intrinsic value - you need it to use the network. Without that there's no reason to hold it and no reason it should be traded for more than 0.

Ethereum had a proposal to allow *contracts* to be responsible for paying fees, so users could send a transaction to a contract with zero ETH fee attached, and the contract could pay fees in ETH to pay miners. So from the user perspective they might be paying fees in some token, but economically the value of ETH is maintained.

Reply:
I would recommend we allow tx fees to be received in any currency by block producers. Reasons:
1) For a user that only wishes to use a third party token (e.g. a stablecoin), the requirement to have a currency they want nothing to do with (i.e. Coda), is terrible UX. Ethereum has faced this problem with its tokens and this has arguably been a major problem preventing its mainstream use as a payment network.
2) Yes, with Ethereum it has been possible to deploy smart contracts where users can pay fees in third party tokens or have gas fees paid by another third party on their behalf (see argent.xyz or https://gsn.openzeppelin.com). However this is possible because 1) Ethereum has smart contract accounts (i.e. non-user controlled accounts), 2) it's Turing complete. Coda has neither of these. Plus, despite these features, this is still not widespread as a design decision in Ethereum.
3) In any given blockchain, its native currency will (at least at first) be the most liquid currency. The decision for a block producer to accept fees in another third party currency is not an easy one if there's not sufficient liquidity for it as no one would be able to pay them. This provides incentives for token developers to develop a liquid market for their token and even provide incentives to block producers to accept their token. This can easily differentiate us from Ethereum and other blockchains used for token settlements.
4) We could still provide exclusivity to Coda for certain uses of our blockchains - e.g. we could limit fees only to Coda tokens for Snark workers. This way we ensure there's always an ongoing demand for Coda tokens.

p.s. I would also argue that tokens don't have 0 value absent a monopoly over tx fees. It would create stronger demand sure, but gold doesn't have 0 value even though gold miners or merchants don't require fees in gold. Its value comes from the global social contract and its provable scarcity (not saying that this should be the only utility of Coda tokens).

Reply:
1. I agree it's bad UX. Ideally conversion would be invisible to the user.
2. I'm 100% fine with making design decisions based on the assumption smart contracts will be in a future version of the protocol. Smart contracts are like half of what makes cryptocurrency interesting.
3. I guess? But see below
4. This seems reasonable, and somewhat allays my concerns.

Gold is actually scarce in a way that cryptocurrency is fundamentally not. There is a finite amount of Au on Earth, and it takes some substantial effort to get at it. Meanwhile I can make as much cryptocurrency as I want simply by forking an existing project, changing one line of source code and recompiling - Bitcoin for example is scarce, cryptocurrencies in general are extremely not. The scarcity is driven solely by social reality. Gold also has intrinsic demand in precision cast parts, as a corrosion resistant coating, and as jewelry.

On a more engineering-y level, allowing transactions with fees paid in non-coda tokens works fine at the consensus layer, but makes transaction gossip much more complicated. Now different block producers have different rules on what transactions they'll accept and forward. We'd need multiple gossip nets, or some other more complicated sync protocol. In addition, a user will see more latency getting transaction where fees are paid in some other token into the blockchain, since fewer block producers are willing to accept them in the first place.

Reply:
Responding to the very last note: this will be a design decision for wallet developers. If a wallet decides there's not enough liquidity as a tx fee token for any other than Coda, they can limit only to Coda tokens. If they decide there's only 1 more stablecoin, they can allow that. If they don't care about simple UX and are a power wallet, they can allow whatever.